src/main/java/org/prebid/server/analytics/reporter/greenbids/GreenbidsAnalyticsReporter.java

```diff
@@ -0,0 +1,389 @@
+                        .build();
+    }
+
+    private static SeatNonBid toSeatNonBid(String bidder, BidRejectionTracker bidRejectionTracker) {
+        final List<NonBid> nonBids = bidRejectionTracker.getRejectionReasons().entrySet().stream()
+                .map(entry -> NonBid.of(entry.getKey(), entry.getValue()))
+                .toList();
+
+        return SeatNonBid.of(bidder, nonBids);
+    }
+
+    private GreenbidsAdUnit createAdUnit(
+            Imp imp,
+            Map<String, Bid> seatsWithBids,
+            Map<String, NonBid> seatsWithNonBids) {
+        final ExtBanner extBanner = getExtBanner(imp.getBanner());
+        final Video video = imp.getVideo();
+        final Native nativeObject = imp.getXNative();
+
+        final MediaTypes mediaTypes = MediaTypes.builder()
+                .banner(extBanner)
+                .video(video)
+                .nativeObject(nativeObject)
+                .build();
+
+        final List<GreenbidsBids> bids = extractBidders(imp, seatsWithBids, seatsWithNonBids);
+
+        final ObjectNode impExt = imp.getExt();
+        final Optional<String> gpidOption = getGpid(impExt);
+        final Optional<String> storedRequestIdOption = getStoredRequestId(impExt);
+        final String adUnitCode = imp.getId();
+
+        final String gpidWithFallback = gpidOption.or(() -> storedRequestIdOption)
+                .orElse(adUnitCode);
+
+        final GreenbidsUnifiedCode.Source codeTypeSource = gpidOption
+                .map(gpid -> GreenbidsUnifiedCode.Source.gpidSource)
+                .or(() -> storedRequestIdOption
+                        .map(storedRequestId -> GreenbidsUnifiedCode.Source.storedRequestIdSource))
+                .orElse(GreenbidsUnifiedCode.Source.adUnitCodeSource);
+
+        final GreenbidsUnifiedCode greenbidsUnifiedCode = GreenbidsUnifiedCode.builder()
+                .value(gpidWithFallback)
+                .source(codeTypeSource)
+                .build();
+
+        return GreenbidsAdUnit.builder()
+                .code(adUnitCode)
+                .unifiedCode(greenbidsUnifiedCode)
+                .mediaTypes(mediaTypes)
+                .bids(bids)
+                .build();
+    }
+
+    public static ExtBanner getExtBanner(Banner banner) {
+        if (banner == null) {
+            return null;
+        }
+
+        final List<List<Integer>> sizes = Optional.ofNullable(banner.getFormat())
+                .filter(formats -> !formats.isEmpty())
+                .map(formats -> formats.stream()
+                        .map(format -> List.of(format.getW(), format.getH()))
+                        .collect(Collectors.toList()))
+                .orElseGet(() -> {
+                    if (banner.getW() != null && banner.getH() != null) {
+                        return List.of(List.of(banner.getW(), banner.getH()));
+                    }
+                    return Collections.emptyList();
+                });
```

Comment:
```
.flatMap(Collection::stream)
.map(format -> List.of(...))
.orElse(banner.getW() != null && banner.getH() != null 
                   ? List.of(List.of(banner.getW(), banner.getH())) 
                   : Collections.emptyList());
```

Reply:
Have added ternary operator 👍 

Concerning flatMap normally the `.flatMap(Collection::stream)` is used to cast `Stream<List<Object>>` to `Stream<Object>`. But in this case we have `Optional<List<Format>>` and inside of Optional we want to access the `List<Format>` and convert it to List<List<Integer>> via passing to `Stream` what we actually do at the moment

```java
final List<List<Integer>> sizes = Optional.ofNullable(banner.getFormat())
                .filter(formats -> !formats.isEmpty())
                .map(formats -> formats.stream()
                        .map(format -> List.of(format.getW(), format.getH()))
                        .collect(Collectors.toList()))
                .orElse(banner.getW() != null && banner.getH() != null
                        ? List.of(List.of(banner.getW(), banner.getH()))
                        : Collections.emptyList());
```